File: benchmarks/python/src/scenarios/sqlalchemy/Scenario.py
import random
from datetime import datetime, timedelta
from sqlalchemy import func, select, update, delete, insert
from sqlalchemy.orm import Session, joinedload
from Connection import get_engine
from scenarios.sqlalchemy.Models import (
    User, Category, Product, Tag, Order, OrderItem, Review
)

"""
SQLAlchemy ORM benchmark scenarios.
Uses Data Mapper pattern via Session.
Models are defined in Models.py to avoid circular imports.
Compared against raw SQL baseline in the same Python runtime.
"""

class Scenario:
    def __init__(self):
        self.engine = get_engine()

    def _session(self) -> Session:
        """Returns a fresh session for each scenario call."""
        return Session(self.engine)

    def a1(self):
        """A1 — Simple select by primary key."""
        with self._session() as session:
            user = session.get(User, random.randint(1, 10000))
            return {
                "id":         user.id,
                "email":      user.email,
                "name":       user.name,
                "created_at": str(user.created_at),
            } if user else {}

    def a2(self):
        """A2 — Filtered list with ORDER BY and LIMIT."""
        with self._session() as session:
            products = session.execute(
                select(Product)
                .where(Product.category_id == random.randint(1, 100))
                .order_by(Product.created_at.desc())
                .limit(20)
            ).scalars().all()

            return [
                {"id": p.id, "name": p.name,
                 "price": str(p.price), "created_at": str(p.created_at)}
                for p in products
            ]

    def a3(self):
        """
        A3 — N+1 diagnostic: load 100 orders then access user for each.
        Intentionally uses lazy loading to trigger N+1 behaviour.
        """
        with self._session() as session:
            orders = session.execute(
                select(Order).order_by(Order.id).limit(100)
            ).scalars().all()

            results = []
            for order in orders:
                # Accessing order.user triggers a separate query (N+1)
                results.append({
                    "order_id": order.id,
                    "total":    str(order.total),
                    "status":   order.status,
                    "user": {
                        "id":    order.user.id,
                        "name":  order.user.name,
                        "email": order.user.email,
                    },
                })
            return results

    def a4(self):
        """A4 — Eager loading: orders with users via joinedload."""
        with self._session() as session:
            orders = session.execute(
                select(Order)
                .options(joinedload(Order.user))
                .order_by(Order.id)
                .limit(100)
            ).scalars().all()

            return [
                {
                    "id":     o.id,
                    "total":  str(o.total),
                    "status": o.status,
                    "user":   {"id": o.user.id, "name": o.user.name},
                }
                for o in orders
            ]

    def b1(self):
        """B1 — Deep eager loading across 3 levels."""
        with self._session() as session:
            order = session.execute(
                select(Order)
                .options(
                    joinedload(Order.items)
                    .joinedload(OrderItem.product)
                    .joinedload(Product.category)
                )
                .where(Order.id == random.randint(1, 200000))
            ).unique().scalar_one_or_none()

            if not order:
                return []

            return [
                {
                    "order_id":     order.id,
                    "total":        str(order.total),
                    "product_name": item.product.name,
                    "category":     item.product.category.name,
                    "quantity":     item.quantity,
                    "price":        str(item.price),
                }
                for item in order.items
            ]

    def b2(self):
        """B2 — Aggregate with GROUP BY."""
        with self._session() as session:
            rows = session.execute(
                select(
                    Category.id,
                    Category.name,
                    func.count(Product.id).label("product_count"),
                    func.round(func.avg(Product.price), 2).label("avg_price"),
                )
                .outerjoin(Product, Product.category_id == Category.id)
                .group_by(Category.id, Category.name)
                .order_by(func.count(Product.id).desc())
            ).all()

            return [
                {"id": r.id, "name": r.name,
                 "product_count": r.product_count,
                 "avg_price": str(r.avg_price)}
                for r in rows
            ]

    def b3(self):
        """B3 — Many-to-many: products by tag with category."""
        with self._session() as session:
            tag = session.get(Tag, random.randint(1, 500))
            if not tag:
                return []

            products = session.execute(
                select(Product)
                .options(joinedload(Product.category))
                .join(Product.tags)
                .where(Tag.id == tag.id)
                .order_by(Product.id)
                .limit(50)
            ).scalars().all()

            return [
                {
                    "id":       p.id,
                    "name":     p.name,
                    "price":    str(p.price),
                    "category": p.category.name,
                }
                for p in products
            ]

    def c1(self):
        """C1 — Bulk insert: 10,000 products in chunks."""
        chunk_size = 500
        total      = 10000
        cat_ids    = list(range(1, 101))
        inserted   = 0

        with self._session() as session:
            for i in range(0, total, chunk_size):
                count = min(chunk_size, total - i)
                rows  = [
                    {
                        "name":        f"Bulk Product {i + j}",
                        "price":       round(random.randint(199, 99999) / 100, 2),
                        "description": None,
                        "category_id": random.choice(cat_ids),
                        "created_at":  datetime.now(),
                    }
                    for j in range(count)
                ]
                session.execute(insert(Product), rows)
                session.commit()
                inserted += count

            # Clean up to restore dataset state
            session.execute(
                delete(Product).where(Product.name.like("Bulk Product%"))
            )
            session.commit()

        return inserted

    def c2(self):
        """C2 — Bulk update: UPDATE FROM subquery, LIMIT 1000 rows.
        SQLAlchemy ORM update does not support LIMIT — use text() with native SQL.
        See methodology/c2_limit_analysis.json for batch size justification.
        """
        from sqlalchemy import text
        thirty_days_ago = datetime.now() - timedelta(days=30)

        with self._session() as session:
            result = session.execute(text("""
                UPDATE orders SET status = 'delivered'
                FROM (
                    SELECT id FROM orders
                    WHERE status = 'shipped'
                      AND created_at < :cutoff
                    ORDER BY id LIMIT 1000
                ) AS batch
                WHERE orders.id = batch.id
            """), {"cutoff": thirty_days_ago})
            session.commit()
            affected = result.rowcount

            # Restore original status
            session.execute(text("""
                UPDATE orders SET status = 'shipped'
                FROM (
                    SELECT id FROM orders
                    WHERE status = 'delivered'
                      AND created_at < :cutoff
                    ORDER BY id LIMIT 1000
                ) AS batch
                WHERE orders.id = batch.id
            """), {"cutoff": thirty_days_ago})
            session.commit()

        return affected

    def d1(self):
        """
        D1 — Unit of Work diagnostic.
        Uses Session flush to demonstrate SQLAlchemy change tracking.
        """
        user_id = random.randint(1, 10000)
        skip    = random.randint(0, 49995)

        with self._session() as session:
            products = session.execute(
                select(Product).order_by(Product.id).limit(5).offset(skip)
            ).scalars().all()

            order = Order(
                user_id=user_id,
                total=0,
                status="pending",
                created_at=datetime.now(),
            )
            session.add(order)
            session.flush()  # Get order.id without committing

            total = 0
            for product in products:
                quantity = random.randint(1, 3)
                price    = float(product.price)
                total   += quantity * price

                item = OrderItem(
                    order_id=order.id,
                    product_id=product.id,
                    quantity=quantity,
                    price=price,
                )
                session.add(item)

            order.total = total
            session.commit()
            order_id = order.id

            # Clean up
            session.execute(
                delete(OrderItem).where(OrderItem.order_id == order_id)
            )
            session.execute(
                delete(Order).where(Order.id == order_id)
            )
            session.commit()

        return {
            "order_id":    order_id,
            "total":       total,
            "items_count": len(products),
        }
    

File: benchmarks/java/src/main/java/com/benchmark/scenarios/hibernate/Scenario.java
package com.benchmark.scenarios.hibernate;
 
import com.benchmark.Connection;
import com.benchmark.scenarios.hibernate.entities.Category;
import com.benchmark.scenarios.hibernate.entities.OrderItem;
import com.benchmark.scenarios.hibernate.entities.Product;
import com.benchmark.scenarios.hibernate.entities.Review;
import com.benchmark.scenarios.hibernate.entities.Tag;
import com.benchmark.scenarios.hibernate.entities.User;
import com.benchmark.scenarios.hibernate.entities.Order;
import jakarta.persistence.EntityManager;
import jakarta.persistence.EntityManagerFactory;
import jakarta.persistence.EntityTransaction;
import org.hibernate.Session;
 
import java.math.BigDecimal;
import java.time.LocalDateTime;
import java.util.*;
 
/**
 * Hibernate ORM benchmark scenarios.
 * Uses Data Mapper pattern via EntityManager and JPA.
 * Second-level cache is disabled for fair benchmarking.
 * Compared against raw SQL baseline in the same Java runtime.
 */
public class Scenario {
 
    private final EntityManagerFactory emf;
 
    public Scenario() {
        this.emf = Connection.getEntityManagerFactory();
    }
 
    /**
     * Returns a fresh EntityManager for each scenario call.
     * First-level cache is cleared between iterations this way.
     */
    private EntityManager em() {
        return emf.createEntityManager();
    }
 
    /**
     * A1 — Simple select by primary key.
     */
    public Map<String, Object> a1() {
        int id = (int) (Math.random() * 10000) + 1;
        try (EntityManager em = em()) {
            User user = em.find(User.class, id);
            if (user == null) return Collections.emptyMap();
            return Map.of(
                "id",         user.getId(),
                "email",      user.getEmail(),
                "name",       user.getName(),
                "created_at", user.getCreatedAt().toString()
            );
        }
    }
 
    /**
     * A2 — Filtered list with ORDER BY and LIMIT.
     */
    public List<Map<String, Object>> a2() {
        int categoryId = (int) (Math.random() * 100) + 1;
        try (EntityManager em = em()) {
            List<Product> products = em.createQuery(
                "SELECT p FROM Product p " +
                "WHERE p.category.id = :catId " +
                "ORDER BY p.createdAt DESC", Product.class)
                .setParameter("catId", categoryId)
                .setMaxResults(20)
                .getResultList();
 
            List<Map<String, Object>> results = new ArrayList<>();
            for (Product p : products) {
                results.add(Map.of(
                    "id",    p.getId(),
                    "name",  p.getName(),
                    "price", p.getPrice()
                ));
            }
            return results;
        }
    }
 
    /**
     * A3 — N+1 diagnostic: load 100 orders then access user for each.
     * Intentionally uses lazy loading to trigger N+1 behaviour.
     */
    public List<Map<String, Object>> a3() {
        try (EntityManager em = em()) {
            List<Order> orders = em.createQuery(
                "SELECT o FROM Order o ORDER BY o.id ASC", Order.class)
                .setMaxResults(100)
                .getResultList();
 
            List<Map<String, Object>> results = new ArrayList<>();
            for (Order order : orders) {
                // Accessing getUser() triggers a separate query (N+1)
                results.add(Map.of(
                    "order_id", order.getId(),
                    "total",    order.getTotal(),
                    "status",   order.getStatus(),
                    "user",     Map.of(
                        "id",    order.getUser().getId(),
                        "name",  order.getUser().getName(),
                        "email", order.getUser().getEmail()
                    )
                ));
            }
            return results;
        }
    }
 
    /**
     * A4 — Eager loading: orders with users via JOIN FETCH.
     */
    public List<Map<String, Object>> a4() {
        try (EntityManager em = em()) {
            List<Order> orders = em.createQuery(
                "SELECT o FROM Order o " +
                "JOIN FETCH o.user u " +
                "ORDER BY o.id ASC", Order.class)
                .setMaxResults(100)
                .getResultList();
 
            List<Map<String, Object>> results = new ArrayList<>();
            for (Order o : orders) {
                results.add(Map.of(
                    "id",     o.getId(),
                    "total",  o.getTotal(),
                    "status", o.getStatus(),
                    "user",   Map.of(
                        "id",   o.getUser().getId(),
                        "name", o.getUser().getName()
                    )
                ));
            }
            return results;
        }
    }
 
    /**
     * B1 — Deep eager loading across 3 levels.
     * Order → OrderItems → Product → Category.
     */
    public Map<String, Object> b1() {
        int orderId = (int) (Math.random() * 200000) + 1;
        try (EntityManager em = em()) {
            Order order = em.createQuery(
                "SELECT o FROM Order o " +
                "LEFT JOIN FETCH o.items i " +
                "LEFT JOIN FETCH i.product p " +
                "LEFT JOIN FETCH p.category " +
                "WHERE o.id = :id", Order.class)
                .setParameter("id", orderId)
                .getResultStream()
                .findFirst()
                .orElse(null);
 
            if (order == null) return Collections.emptyMap();
 
            List<Map<String, Object>> items = new ArrayList<>();
            for (OrderItem item : order.getItems()) {
                items.add(Map.of(
                    "product",  item.getProduct().getName(),
                    "category", item.getProduct().getCategory().getName(),
                    "quantity", item.getQuantity(),
                    "price",    item.getPrice()
                ));
            }
 
            return Map.of(
                "order_id", order.getId(),
                "total",    order.getTotal(),
                "items",    items
            );
        }
    }
 
    /**
     * B2 — Aggregate with GROUP BY using JPQL.
     */
    public List<Map<String, Object>> b2() {
        try (EntityManager em = em()) {
            List<Object[]> rows = em.createQuery(
                "SELECT c.id, c.name, COUNT(p.id), AVG(p.price) " +
                "FROM Category c LEFT JOIN c.products p " +
                "GROUP BY c.id, c.name " +
                "ORDER BY COUNT(p.id) DESC", Object[].class)
                .getResultList();
 
            List<Map<String, Object>> results = new ArrayList<>();
            for (Object[] row : rows) {
                results.add(Map.of(
                    "id",            row[0],
                    "name",          row[1],
                    "product_count", row[2],
                    "avg_price",     row[3] != null ? row[3] : 0.0
                ));
            }
            return results;
        }
    }
 
    /**
     * B3 — Many-to-many: products by tag with category.
     */
    public List<Map<String, Object>> b3() {
        int tagId = (int) (Math.random() * 500) + 1;
        try (EntityManager em = em()) {
            List<Product> products = em.createQuery(
                "SELECT p FROM Product p " +
                "JOIN FETCH p.category " +
                "JOIN p.tags t " +
                "WHERE t.id = :tagId " +
                "ORDER BY p.id ASC", Product.class)
                .setParameter("tagId", tagId)
                .setMaxResults(50)
                .getResultList();
 
            List<Map<String, Object>> results = new ArrayList<>();
            for (Product p : products) {
                results.add(Map.of(
                    "id",       p.getId(),
                    "name",     p.getName(),
                    "price",    p.getPrice(),
                    "category", p.getCategory().getName()
                ));
            }
            return results;
        }
    }
 
    /**
     * C1 — Bulk insert: 10,000 products using JDBC batch via Hibernate session.
     */
    public int c1() {
        int chunkSize = 500;
        int total     = 10000;
        int[] catIds  = new int[100];
        for (int i = 0; i < 100; i++) catIds[i] = i + 1;
 
        try (EntityManager em = em()) {
            EntityTransaction tx = em.getTransaction();
            tx.begin();
 
            Session session = em.unwrap(Session.class);
            session.doWork(conn -> {
                try (var ps = conn.prepareStatement(
                        "INSERT INTO products (name, price, category_id, created_at) " +
                        "VALUES (?, ?, ?, NOW())")) {
                    for (int i = 0; i < total; i++) {
                        ps.setString(1, "Bulk Product " + i);
                        ps.setBigDecimal(2, BigDecimal.valueOf(
                            Math.round(Math.random() * 99800 + 199) / 100.0));
                        ps.setInt(3, catIds[(int)(Math.random() * 100)]);
                        ps.addBatch();
                        if ((i + 1) % chunkSize == 0) ps.executeBatch();
                    }
                    ps.executeBatch();
                }
            });
 
            tx.commit();
 
            // Clean up
            tx.begin();
            em.createQuery(
                "DELETE FROM Product p WHERE p.name LIKE 'Bulk Product%'")
                .executeUpdate();
            tx.commit();
        }
 
        return total;
    }
 
    /**
     * C2 — Bulk update using JPQL UPDATE for performance.
     */
    public int c2() {
        LocalDateTime thirtyDaysAgo = LocalDateTime.now().minusDays(30);
 
        try (EntityManager em = em()) {
            EntityTransaction tx = em.getTransaction();
            tx.begin();
 
            // JPQL does not support LIMIT on UPDATE — use native SQL with UPDATE FROM subquery
            int affected = (int) em.createNativeQuery(
                "UPDATE orders SET status = 'delivered' " +
                "FROM (" +
                "  SELECT id FROM orders " +
                "  WHERE status = 'shipped' " +
                "  AND created_at < :cutoff " +
                "  ORDER BY id LIMIT 1000" +
                ") AS batch WHERE orders.id = batch.id")
                .setParameter("cutoff", thirtyDaysAgo)
                .executeUpdate();

            tx.commit();

            // Restore original status
            tx.begin();
            em.createNativeQuery(
                "UPDATE orders SET status = 'shipped' " +
                "FROM (" +
                "  SELECT id FROM orders " +
                "  WHERE status = 'delivered' " +
                "  AND created_at < :cutoff " +
                "  ORDER BY id LIMIT 1000" +
                ") AS batch WHERE orders.id = batch.id")
                .setParameter("cutoff", thirtyDaysAgo)
                .executeUpdate();
            tx.commit();
 
            return affected;
        }
    }
 
    /**
     * D1 — Unit of Work diagnostic.
     * Uses EntityManager persist/flush to demonstrate Hibernate's
     * change tracking and deferred write behaviour.
     */
    public Map<String, Object> d1() {
        int userId = (int) (Math.random() * 10000) + 1;
        int skip   = (int) (Math.random() * 49995);
 
        try (EntityManager em = em()) {
            EntityTransaction tx = em.getTransaction();
            tx.begin();
 
            User user = em.find(User.class, userId);
 
            List<Product> products = em.createQuery(
                "SELECT p FROM Product p ORDER BY p.id ASC",
                Product.class)
                .setFirstResult(skip)
                .setMaxResults(5)
                .getResultList();
 
            Order order = new Order();
            order.setUser(user);
            order.setTotal(BigDecimal.ZERO);
            order.setStatus("pending");
            order.setCreatedAt(LocalDateTime.now());
            em.persist(order);
            em.flush();
 
            BigDecimal total = BigDecimal.ZERO;
            List<OrderItem> items = new ArrayList<>();
 
            for (Product product : products) {
                int quantity = (int)(Math.random() * 3) + 1;
 
                OrderItem item = new OrderItem();
                item.setOrder(order);
                item.setProduct(product);
                item.setQuantity(quantity);
                item.setPrice(product.getPrice());
                em.persist(item);
                items.add(item);
 
                total = total.add(
                    product.getPrice().multiply(BigDecimal.valueOf(quantity))
                );
            }
 
            order.setTotal(total);
            em.flush();
            tx.commit();
 
            int orderId = order.getId();
 
            // Clean up
            tx.begin();
            em.createQuery(
                "DELETE FROM OrderItem i WHERE i.order.id = :id")
                .setParameter("id", orderId)
                .executeUpdate();
            em.createQuery(
                "DELETE FROM Order o WHERE o.id = :id")
                .setParameter("id", orderId)
                .executeUpdate();
            tx.commit();
 
            return Map.of(
                "order_id",    orderId,
                "total",       total,
                "items_count", items.size()
            );
        }
    }
}
